File: tests/refs/dijkstra_shortest_path.cpp
// Reference C++ driver for tests/cases/dijkstra_shortest_path.json.
// Undirected graph; unreachable -> -1.
#include <climits>
#include <iostream>
#include <queue>
#include <sstream>
#include <string>
#include <vector>

std::vector<long long> dijkstra(int V, const std::vector<std::tuple<int,int,long long>>& edges, int src) {
    std::vector<std::vector<std::pair<int,long long>>> adj(V);
    for (auto& [u, v, w] : edges) { adj[u].push_back({v, w}); adj[v].push_back({u, w}); }
    const long long INF = LLONG_MAX / 2;
    std::vector<long long> dist(V, INF);
    dist[src] = 0;
    using P = std::pair<long long, int>;
    std::priority_queue<P, std::vector<P>, std::greater<P>> pq;
    pq.push({0, src});
    while (!pq.empty()) {
        auto [d, u] = pq.top(); pq.pop();
        if (d > dist[u]) continue;
        for (auto& [v, w] : adj[u]) {
            long long nd = d + w;
            if (nd < dist[v]) { dist[v] = nd; pq.push({nd, v}); }
        }
    }
    std::vector<long long> out(V);
    for (int i = 0; i < V; ++i) out[i] = (dist[i] == INF) ? -1 : dist[i];
    return out;
}

int main() {
    std::string line, name;
    int V = 0, src = 0, edgesRemain = 0; bool inEdges = false;
    std::vector<std::tuple<int,int,long long>> edges;
    std::vector<long long> expected;
    bool hi = false, he = false; int p = 0, f = 0;
    auto run = [&]() {
        if (name.empty() || !hi || !he) return;
        auto got = dijkstra(V, edges, src);
        if (got == expected) { std::cout << "PASS dijkstra_shortest_path :: " << name << "\n"; ++p; }
        else { std::cout << "FAIL dijkstra_shortest_path :: " << name << "\n"; ++f; }
        name.clear(); edges.clear(); expected.clear();
        hi = he = false; inEdges = false;
    };
    while (std::getline(std::cin, line)) {
        if (line.empty()) { run(); continue; }
        std::istringstream iss(line); std::string tag; iss >> tag;
        if (tag == "CASE") { iss >> name; }
        else if (tag == "V") { iss >> V; }
        else if (tag == "EDGES") { iss >> edgesRemain; inEdges = true; edges.clear(); if (edgesRemain == 0) inEdges = false; }
        else if (tag == "SRC") { iss >> src; hi = true; }
        else if (tag == "ARR") {
            std::size_t n; iss >> n;
            expected.clear();
            for (std::size_t i = 0; i < n; ++i) { long long v; iss >> v; expected.push_back(v); }
            he = true;
        }
        else if (inEdges) {
            std::istringstream is2(line); int u, v; long long w; is2 >> u >> v >> w;
            edges.push_back({u, v, w}); --edgesRemain;
            if (edgesRemain == 0) inEdges = false;
        }
    }
    run();
    std::cout << "TOTAL: " << p << " passed, " << f << " failed\n";
    return f == 0 ? 0 : 1;
}

File: tests/refs/dijkstra_shortest_path.rs
// Reference Rust driver for tests/cases/dijkstra_shortest_path.json.
use std::collections::BinaryHeap;
use std::cmp::Reverse;
use std::io::{self, BufRead};

fn dijkstra(v: usize, edges: &[(i32, i32, i64)], src: usize) -> Vec<i64> {
    let mut adj: Vec<Vec<(usize, i64)>> = vec![Vec::new(); v];
    for &(u, w, ww) in edges {
        adj[u as usize].push((w as usize, ww));
        adj[w as usize].push((u as usize, ww));
    }
    let inf = i64::MAX / 2;
    let mut dist = vec![inf; v];
    dist[src] = 0;
    let mut pq: BinaryHeap<Reverse<(i64, usize)>> = BinaryHeap::new();
    pq.push(Reverse((0, src)));
    while let Some(Reverse((d, u))) = pq.pop() {
        if d > dist[u] { continue; }
        for &(x, ww) in &adj[u] {
            let nd = d + ww;
            if nd < dist[x] { dist[x] = nd; pq.push(Reverse((nd, x))); }
        }
    }
    dist.iter().map(|&d| if d == inf { -1 } else { d }).collect()
}

fn main() {
    let stdin = io::stdin();
    let mut name = String::new();
    let mut v_in: usize = 0; let mut src: usize = 0;
    let mut edges: Vec<(i32, i32, i64)> = Vec::new();
    let mut edges_remain: usize = 0; let mut in_edges = false;
    let mut expected: Vec<i64> = Vec::new();
    let mut hi = false; let mut he = false;
    let mut p = 0u32; let mut f = 0u32;
    let flush = |name: &mut String, v_in: usize, edges: &mut Vec<(i32, i32, i64)>, src: usize, expected: &mut Vec<i64>,
                 hi: &mut bool, he: &mut bool, in_edges: &mut bool, p: &mut u32, f: &mut u32| {
        if !name.is_empty() && *hi && *he {
            let got = dijkstra(v_in, edges, src);
            if got == *expected { println!("PASS dijkstra_shortest_path :: {}", name); *p += 1; }
            else { println!("FAIL dijkstra_shortest_path :: {}\n  expected: {:?}\n  got: {:?}", name, expected, got); *f += 1; }
        }
        name.clear(); edges.clear(); expected.clear();
        *hi = false; *he = false; *in_edges = false;
    };
    for line in stdin.lock().lines() {
        let line = line.unwrap_or_default();
        if line.is_empty() { flush(&mut name, v_in, &mut edges, src, &mut expected, &mut hi, &mut he, &mut in_edges, &mut p, &mut f); continue; }
        let mut parts = line.split_whitespace();
        let tag = parts.next().unwrap_or("");
        match tag {
            "CASE" => { name = parts.next().unwrap_or("").to_string(); }
            "V" => { v_in = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0); }
            "EDGES" => {
                edges_remain = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                in_edges = true; edges.clear();
                if edges_remain == 0 { in_edges = false; }
            }
            "SRC" => { src = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0); hi = true; }
            "ARR" => {
                let n: usize = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                expected.clear();
                for _ in 0..n { if let Some(t) = parts.next() { if let Ok(v) = t.parse::<i64>() { expected.push(v); } } }
                he = true;
            }
            _ => {
                if in_edges {
                    let u: i32 = tag.parse().unwrap_or(0);
                    let w: i32 = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                    let ww: i64 = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                    edges.push((u, w, ww));
                    edges_remain -= 1;
                    if edges_remain == 0 { in_edges = false; }
                }
            }
        }
    }
    flush(&mut name, v_in, &mut edges, src, &mut expected, &mut hi, &mut he, &mut in_edges, &mut p, &mut f);
    println!("TOTAL: {} passed, {} failed", p, f);
    std::process::exit(if f == 0 { 0 } else { 1 });
}
